vllm_spyre/multimodal/mm_mappings/llava_next.py

```diff
@@ -0,0 +1,158 @@
+    def get_mm_specific_load_overrides(hf_config: PretrainedConfig):
+        """Get any overrides needed for initializing the FMS model from the
+        transformers config. For this model, we need to fix the head_dim, which
+        currently surfaces as a problem for all 2b variants of granite 3.x LLMs
+        when running through FMS.
+        
+        TODO: If additional variants of granite vision are added, or broader
+        llava next support is added in FMS, handle it properly here.
+        """
+        serialization.extend_adapter(
+            "llava_next", "hf", ["weight_expansion_for_mismatched_head_dim"])
+        return {
+            "override_hf_pretrained_config": True,
+            "text_config": {
+                "head_dim": 128
+            },
+        }
+
+    @staticmethod
+    def get_maybe_mm_embeddings(
+        fms_model: torch.nn.Module,
+        input_ids: torch.Tensor,
+        mm_features: list[MultiModalFeatureSpec],
+        is_decode: bool,
+    ) -> torch.Tensor:
+        """Get the text or multimodal embeddings for Llava Next using
+        the (potentially compiled) FMS model.
+        """
+        fms_kwargs = {"use_cache": True}
+        mm_spec_keys = ["pixel_values", "image_sizes"]
+
+        # Only merge multimodal features in prefill; nothing mm in decode
+        if mm_features:
+            if len(mm_features) != 1:
+                raise ValueError(
+                    "Currently we assume we only embed one mm request at a time"
+                )
+            mm_spec = mm_features[0].data
+            if mm_spec is not None:
+                # NOTE: This should be pretty safe as it's dependent on the
+                # vLLM/HF processor objects, but we check it anyway to be safe
+                # for now, since transformers 5.0 is just around the corner.
+                if any(k not in mm_spec for k in mm_spec_keys):
+                    raise KeyError(
+                        f"Llava Next requires kwargs: {mm_spec_keys}")
+
+                fms_kwargs["pixel_values"] = mm_spec["pixel_values"].data
+                image_sizes = mm_spec["image_sizes"].data
+
+                # Careful about this; if it's 1D, we'll a tensor of shape
+                # [x, y], which will break in a weird way in image packing,
+                # since it assumes it's 2D and will get sad about getting
+                # an int instead of an iterable
+                if image_sizes.ndim == 1:
+                    image_sizes = image_sizes.unsqueeze(0)
+                fms_kwargs["image_sizes"] = image_sizes
+
+        # NOTE: use_cache is actually not used here, but currently it's
+        # required. Also, the value of iteration for decode as long as it's > 0.
+        input_embeds, _ = fms_model.prepare_inputs_for_generation(
+            iteration=0 if not is_decode else 1,
+            input_ids=input_ids,
+            kwargs=fms_kwargs)
+        return input_embeds
+
+    def get_warmup_mm_features(self):
```

Comment:
Is it possible to re-use the same vllm code path for creating mm_features out of multimodal data instead of requiring hand-rolled warmup mm_features per model type?

It seems like there's room for bugs here if the features aren't built correctly, and the possibility of code drift between how features are created here vs. at inference time